### PR TITLE
Refactored amcl code for modularization of motion models(main branch)

### PR DESCRIPTION
**Pull request by poornimajd**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | add tickets here #2552 |
| Primary OS tested on | Ubuntu 20.04 |
| Robotic platform tested on | turtlebot3 |

---

## Description of contribution in a few bullet points
- The amcl source code has been slightly modified and a plugin package for motion model has been created.

- Pluginlib is used to make the differential and omni model as plugins.Major files changed are related to the motion_model folder.

- This functionality will enable a developer to test his/her own custom motion model by just creating it as a plugin and adding the name in the nav2_params.yaml file.
-  Code is tested on the ros2 foxy branch.

## Description of documentation updates required from your changes

- The parameter robot_model_type in nav2_params.yaml file is to be taken as the fully-qualified type of the plugin class.For example - robot_model_type: "nav2_amcl::DifferentialMotionModel" if the name of the custom motion model class is "DifferentialMotionModel" and the namespace is "nav2_amcl".

- To test any custom motion model ,just add the motion_model.hpp file to the source code of the motion model implemented as a plugin.

---
#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Comment by poornimajd** [comment]
Quick update- I have tested the navigation2 main branch with the refactored amcl code(this PR) locally,and I did not get any errors i.e The code was tested on ros2 rolling distribution in a docker container.Ubuntu(20.04).

**Comment by SteveMacenski** [comment]
OK, try pulling in main or rebasing so you have the latest status of the main branch, maybe its an outdated thing?

File: nav2_bringup/params/nav2_params.yaml [hunk 1]
```diff
@@ -26,7 +26,7 @@ amcl:
     recovery_alpha_fast: 0.0
     recovery_alpha_slow: 0.0
     resample_interval: 1
-    robot_model_type: "differential"
+    robot_model_type: "nav2_amcl::DifferentialMotionModel"
```
Comment (SteveMacenski) [inline nav2_bringup/params/nav2_params.yaml:29]: There are other yaml files in this directory that would need to be updated as well (and in the `nav2_system_tests`, try searching for `robot_model_type` and replacing with this)

**Review by SteveMacenski — COMMENTED** [review]
One last change here and overall LGTM when we can get this turning over in CI. 

You must also add in an entry into this page https://github.com/ros-planning/navigation.ros.org/blob/master/migration/Galactic.rst for the migration guide that this has changed

File: nav2_bringup/params/nav2_params.yaml (hunk 1 above)
Comment (SteveMacenski) [reply nav2_bringup/params/nav2_params.yaml:29]: ohhhh that actually might be why your tests are failing!

Comment (poornimajd) [reply nav2_bringup/params/nav2_params.yaml:29]: Changed the robot_model_type in the yaml files wherever required.

**Comment by poornimajd** [comment]
> OK, try pulling in main or rebasing so you have the latest status of the main branch, maybe its an outdated thing?

Thank you for the suggestion.But I tried to test it again with the updated and latest main branch on the local system,and it shows no errors.Also the PR is in sync with the updated main branch.

**Comment by SteveMacenski** [comment]
OK, these test failures now look normal with a couple of flaky tests, I don't see anything concerning in CI anymore

**Comment by SteveMacenski** [comment]
Docs PR https://github.com/ros-planning/navigation.ros.org/pull/231#pullrequestreview-803032177

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
Did another look, this is good to merge once the docs PR is good to go, just need to make 2 really small updates and this is all done

**Comment by SteveMacenski** [comment]
Thanks for the contribution and thanks for putting up with my bunch of change requests :laughing:

**Comment by poornimajd** [comment]
Thank you for the helpful suggestions.

**Comment by SteveMacenski** [comment]
FYI @poornimajd https://github.com/ros-planning/navigation2/issues/2673 your PR has a few issues, if you had a few minutes to see these errors and address them

**Comment by poornimajd** [comment]
I will have a look at it and try to resolve it.

**Pull request merged**